src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs

```diff
@@ -376,18 +383,98 @@ internal void DoGetExtensionMethods(ArrayBuilder<MethodSymbol> methods, string n
         }
 
 #nullable enable
+        private static bool IsValidExtensionReceiverParameter(ParameterSymbol thisParam)
+        {
+            Debug.Assert(thisParam is not null);
+
+            // For ref and ref-readonly extension members and classic extension methods, receivers need to be of the correct types to be considered in lookup
+            if (thisParam.RefKind == RefKind.Ref && !thisParam.Type.IsValueType)
+            {
+                return false;
+            }
 
-        internal void GetExtensionContainers(ArrayBuilder<NamedTypeSymbol> extensions)
+            if (thisParam.RefKind is RefKind.In or RefKind.RefReadOnlyParameter && thisParam.Type.TypeKind != TypeKind.Struct)
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        internal void GetExtensionMembers(ArrayBuilder<Symbol> members, string? name, string? alternativeName, int arity, LookupOptions options)
```

Comment:
>GetExtensionMembers [](http://example.com/codeflow?start=22&length=19)

It might be worth adding a comment saying that this method doesn't perform complete lookup viability check (like accessibility, etc.) And for other APIs that take LookupOptions with the only purpose to pass it to this method as well.